timApp/tide_support/routes.py

```diff
@@ -0,0 +1,126 @@
+from authlib.integrations.flask_oauth2 import current_token
+from flask import Response, request
+
+from timApp.auth.oauth2.models import Scope
+from timApp.auth.oauth2.oauth2 import require_oauth
+
+from timApp.tide_support.utils import (
+    user_ide_courses,
+    ide_tasks,
+    ide_task_by_id,
+    ide_task_folders_by_doc,
+    ide_submit_task,
+)
+from timApp.user.user import User
+from timApp.util.flask.responsehelper import json_response
+from timApp.util.flask.typedblueprint import TypedBlueprint
+
+ide = TypedBlueprint("ide", __name__, url_prefix="/ide")
+
+
+@ide.get("ideCourses")
+@require_oauth(Scope.user_tasks.value)
+def get_user_ide_courses() -> Response:
+    """
+    Get all courses that the user has bookmarked and have ideCourse tag
+    :return: JSON response with all courses and their task folders
+    """
+
+    user: User = current_token.user
+    return json_response(user_ide_courses(user=user))
+
+
+@ide.get("taskFoldersByDoc")
+@require_oauth(Scope.user_tasks.value)
+def get_ide_task_folders_by_doc() -> Response:
+    """
+    Get all ide tasks folders by document id
+    :return: JSON response with the list of folders
+    """
+
+    doc_id = request.json.get("doc_id")
+    doc_path = request.json.get("doc_path")
+
+    if not doc_id and not doc_path:
+        return json_response({"error": "No doc_id or doc_path provided"})
+
+    return json_response(ide_task_folders_by_doc(doc_id=doc_id, doc_path=doc_path))
+
+
+@ide.get("tasksByDoc")
+@require_oauth(Scope.user_tasks.value)
+def get_ide_tasks_by_doc_path() -> Response:
+    """
+    Get all tasks by task folder path or doc_id
+    :return: JSON response with the task
+    """
+    doc_path = request.json.get("doc_path")
+    doc_id = request.json.get("doc_id")
```

Comment:
Meillä polkuja varten on "erikoinen" [TypedBlueprint](https://tim.jyu.fi/view/tim/TIMin-kehitys/Johdatus-TIMin-kehitykseen#flaskin-reitit), jonka avulla kaikki nuo request.json yms voi laittaa suoraan polkufunktion parametrikisi. Eli siis tuollainen on mahdollinen ja sallittu

```suggestion
@oauth.get("demos-by-doc-id")
@require_oauth(Scope.user_tasks.value)
def get_ide_demos_by_doc_id(doc_id: int) -> Response:
    """
    Get all demos by document id
    :return: JSON response with the task
    """
```

Tämän hyödyt:

- doc_id on nyt pakko olla int
- doc_id on nyt pakko olla määritelty (se voi olla vapaaehtoinen, jos antaa argumentille oletusarvon)

"Haitat":

- GET-poluissa argumentit otetaan URL-parametrista (eli siis `/oauth/demos-by-doc-id?doc_id=1`) eikä se vielä tue arvon nappaamista JSONista
- POST-polussa argumentit otetaan JSONista eikä se vielä tue arvon nappaamista URL-parametrista

Sanoisin, että kuitenkin tuo on varsin kätevä tapa määritellä polulle parametreja.